File: backend/bebcare/schemas/generation_history.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class GenerationHistoryUserSummary(BaseModel):
    user_id: str
    username: str
    email: str


class GenerationHistoryProductSummary(BaseModel):
    product_id: str
    name: str


class GenerationHistoryQaSummary(BaseModel):
    hard_fail_count: int = 0
    warning_count: int = 0


class GenerationHistoryListItem(BaseModel):
    run_id: str
    created_at: datetime
    status: str
    source: str
    user: GenerationHistoryUserSummary
    product: Optional[GenerationHistoryProductSummary] = None
    thumbnail_url: Optional[str] = None
    credits_charged: int = 0
    latency_ms: Optional[int] = None
    provider_summary: Optional[str] = None
    qa_summary: GenerationHistoryQaSummary = Field(default_factory=GenerationHistoryQaSummary)
    diagnosis_line: str = ""


class GenerationHistoryListResponse(BaseModel):
    items: list[GenerationHistoryListItem]
    total: int
    page: int
    page_size: int
    pages: int


class GenerationHistoryArtifact(BaseModel):
    artifact_id: str
    cdn_url: str
    selected: bool
    persistence_warning: Optional[str] = None
    candidate_index: int


class GenerationHistoryFinding(BaseModel):
    finding_id: str
    stage: str
    check_code: str
    check_label: str
    severity: str
    passed: bool
    details: Optional[dict[str, Any]] = None
    qa_kind: str = "deterministic"
    confidence: Optional[str] = None
    created_at: datetime


class GenerationHistoryCreditInfo(BaseModel):
    charged: int = 0
    reservation_status: Optional[str] = None
    grant_source: Optional[str] = None
    grant_note: Optional[str] = None


class GenerationHistoryGenerateTask(BaseModel):
    task_id: str
    status: str
    progress: int = 0
    stage: Optional[str] = None
    result: Optional[dict[str, Any]] = None
    expired: bool = False


class GenerationHistoryCompareSibling(BaseModel):
    run_id: str
    status: str
    image_prompt_pipeline: Optional[str] = None
    thumbnail_url: Optional[str] = None
    selected: bool = False
    created_at: datetime


class GenerationHistoryDetail(BaseModel):
    run_id: str
    owner_user_id: str
    source: str
    product_id: Optional[str] = None
    scheduled_task_id: Optional[str] = None
    generate_task_id: Optional[str] = None
    rollout_mode_at_start: str
    experiment_variant: Optional[str] = None
    requested_pipeline_version: str
    executed_pipeline_version: str
    fallback_reason: Optional[str] = None
    fallback_path: Optional[str] = None
    image_prompt_pipeline: Optional[str] = None
    compare_group_id: Optional[str] = None
    provider_type: Optional[str] = None
    provider_id: Optional[str] = None
    model: Optional[str] = None
    image_size: Optional[str] = None
    image_provider_mode: Optional[str] = None
    status: str
    error_category: Optional[str] = None
    latency_ms: Optional[int] = None
    retry_count: Optional[int] = None
    credits_charged: int = 0
    provider_usage: Optional[dict[str, Any]] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    quality_protection_mode: Optional[str] = None
    quality_policy_version: Optional[str] = None
    product_fidelity_prevention_mode: Optional[str] = None
    visual_fidelity_qa_mode: Optional[str] = None
    visual_fidelity_policy_version: Optional[str] = None
    requested_selector_strategy: Optional[str] = None
    executed_selector_strategy: Optional[str] = None
    selection_seed: Optional[str] = None
    user: GenerationHistoryUserSummary
    product: Optional[GenerationHistoryProductSummary] = None
    output_snapshot: Optional[dict[str, Any]] = None
    generate_task: Optional[GenerationHistoryGenerateTask] = None
    artifacts: list[GenerationHistoryArtifact] = Field(default_factory=list)
    quality_findings: list[GenerationHistoryFinding] = Field(default_factory=list)
    credit: GenerationHistoryCreditInfo = Field(default_factory=GenerationHistoryCreditInfo)
    generation_plan: Optional[dict[str, Any]] = None
    reference_manifest: Optional[dict[str, Any]] = None
    qa_summary: GenerationHistoryQaSummary = Field(default_factory=GenerationHistoryQaSummary)
    diagnosis_line: str = ""
    compare_siblings: list[GenerationHistoryCompareSibling] = Field(default_factory=list)
    generation_diagnostics: Optional[dict[str, Any]] = None

File: frontend/src/api/generationHistory.ts
import axiosInstance from './axiosInstance';
import type { ExecutionDimensions } from './tasks';

export interface GenerationHistoryUser {
  user_id: string;
  username: string;
  email: string;
}

export interface GenerationHistoryProduct {
  product_id: string;
  name: string;
}

export interface GenerationHistoryQaSummary {
  hard_fail_count: number;
  warning_count: number;
}

export interface GenerationHistoryListItem {
  run_id: string;
  created_at: string;
  status: string;
  source: string;
  user: GenerationHistoryUser;
  product?: GenerationHistoryProduct | null;
  thumbnail_url?: string | null;
  credits_charged: number;
  latency_ms?: number | null;
  provider_summary?: string | null;
  qa_summary: GenerationHistoryQaSummary;
  diagnosis_line: string;
}

export interface GenerationHistoryListResponse {
  items: GenerationHistoryListItem[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface GenerationHistoryArtifact {
  artifact_id: string;
  cdn_url: string;
  selected: boolean;
  persistence_warning?: string | null;
  candidate_index: number;
}

export interface GenerationHistoryFinding {
  finding_id: string;
  stage: string;
  check_code: string;
  check_label: string;
  severity: string;
  passed: boolean;
  details?: Record<string, unknown> | null;
  qa_kind: string;
  confidence?: string | null;
  created_at: string;
}

export interface GenerationHistoryCreditInfo {
  charged: number;
  reservation_status?: string | null;
  grant_source?: string | null;
  grant_note?: string | null;
}

export interface GenerationHistoryGenerateTask {
  task_id: string;
  status: string;
  progress: number;
  stage?: string | null;
  result?: Record<string, unknown> | null;
  expired: boolean;
}

export interface GenerationHistoryCompareSibling {
  run_id: string;
  status: string;
  image_prompt_pipeline?: string | null;
  thumbnail_url?: string | null;
  selected: boolean;
  created_at: string;
}

export interface GenerationHistoryOutputSnapshot {
  image_prompt?: string;
  copywriting?: string;
  dimensions?: ExecutionDimensions;
  warning?: string;
  error?: string;
  reference_product_images?: string[];
  reference_scene_images?: string[];
}

export interface GenerationHistoryDetail {
  run_id: string;
  owner_user_id: string;
  source: string;
  product_id?: string | null;
  scheduled_task_id?: string | null;
  generate_task_id?: string | null;
  rollout_mode_at_start: string;
  experiment_variant?: string | null;
  requested_pipeline_version: string;
  executed_pipeline_version: string;
  fallback_reason?: string | null;
  fallback_path?: string | null;
  image_prompt_pipeline?: string | null;
  compare_group_id?: string | null;
  provider_type?: string | null;
  provider_id?: string | null;
  model?: string | null;
  image_size?: string | null;
  image_provider_mode?: string | null;
  status: string;
  error_category?: string | null;
  latency_ms?: number | null;
  retry_count?: number | null;
  credits_charged: number;
  provider_usage?: Record<string, unknown> | null;
  created_at: string;
  completed_at?: string | null;
  quality_protection_mode?: string | null;
  quality_policy_version?: string | null;
  product_fidelity_prevention_mode?: string | null;
  visual_fidelity_qa_mode?: string | null;
  visual_fidelity_policy_version?: string | null;
  requested_selector_strategy?: string | null;
  executed_selector_strategy?: string | null;
  selection_seed?: string | null;
  user: GenerationHistoryUser;
  product?: GenerationHistoryProduct | null;
  output_snapshot?: GenerationHistoryOutputSnapshot | null;
  generate_task?: GenerationHistoryGenerateTask | null;
  artifacts: GenerationHistoryArtifact[];
  quality_findings: GenerationHistoryFinding[];
  credit: GenerationHistoryCreditInfo;
  generation_plan?: Record<string, unknown> | null;
  reference_manifest?: Record<string, unknown> | null;
  qa_summary: GenerationHistoryQaSummary;
  diagnosis_line: string;
  compare_siblings: GenerationHistoryCompareSibling[];
}

export interface GenerationHistoryListParams {
  page?: number;
  page_size?: number;
  user_id?: string;
  username?: string;
  email?: string;
  product_id?: string;
  status?: string;
  source?: string;
  date_from?: string;
  date_to?: string;
  has_qa_failures?: boolean;
  credits_charged_min?: number;
}

export const listGenerationRuns = async (
  params: GenerationHistoryListParams = {},
): Promise<GenerationHistoryListResponse> => {
  const response = await axiosInstance.get('/admin/generation-runs', { params });
  const data = response.data;
  return {
    items: Array.isArray(data?.items) ? data.items : [],
    total: data?.total ?? 0,
    page: data?.page ?? 1,
    page_size: data?.page_size ?? 20,
    pages: data?.pages ?? 1,
  };
};

export const getGenerationRun = async (runId: string): Promise<GenerationHistoryDetail> => {
  const response = await axiosInstance.get(`/admin/generation-runs/${runId}`);
  return response.data;
};
